### PR TITLE
Add options to control if ".uid" files are generated

### LINKED ISSUE #11571: Problem with UIDs
### Describe the project you are working on

N/A

### Describe the problem or limitation you are having in your project

With https://godotengine.org/article/uid-changes-coming-to-godot-4-4/ I believe UIDs might make some problems for people, especially ones who forget to copy UID file.
Also I believe some people prefer to have to update paths in code, if there is need to do so. 
Also how many times are files being moved across project?
I feel like for some assets or scripts once you properly categorize them you will never move them anywhere else.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Option to disable generation of UID files.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

When converting projects to 4.4 godot would ask do you want to enable UID files.
When creating new project using godot 4.4 it would have option does dev want to use UID files.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

N/A

### Is there a reason why this should be core and not an add-on in the asset library?

N/A

### PR DESCRIPTION
**Pull request by Daylily-Zeleen**
The main purpose of this pr is to allow not generate ".uid" files for "res://addons/", by setting "editor/resource/generate_uid_files_for_addons" to false in project settings.

Sometimes plugins are not included in version control system( for example, using [gd-plug](https://github.com/imjp94/gd-plug) to reference plugins by single script), generate ".uid" for plugin's scripts and reference them by using uid are not reasonable.

Additionally, I add "editor/resource/generate_uid_files" to allow not generate ".uid" for the project, but default is true.
Generate ".uid" for each script make the file resource manager become ugly, and we should manipulate 2 files now.
There are some personal preferences here, I believe some people also have this feeling.
For old users, they should used drag files in editor's file system dock, it should not breaks the reference relationship by using this way, and uid is unreadable, we don't use them for coding directly.
In other words, ".uid" files are not necessary for some users.

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/11571*

### CODE REVIEW

**Comment by JoNax97** [comment]
I think Godot should commit one way or the other. Leaving this as an optional feature greatly limits it's potential in the future. All future features that build up from the uid system (or the metadata-companion-file concept in general) will be impossible to implement fully if some users don't have uids available.

**Comment by Daylily-Zeleen** [comment]
@JoNax97 What features? You should provide specific situation to make your viewpoint convincing and solve the problem that I mentioned about (".uid" in "res://addons").

If you want to talk about future, to make plugins' detail files outside the version control system is a good future, too.
To be honest, plugins should be referenced like nuget in C# instead of including its detail files, right?

**Comment by JoNax97** [comment]
I agree with you that add-ons should be handled via some package management system. But until them we cannot guarantee that a plugin does or does not rely on uids and so it's not safe to disables them for all addons. 

My main concern is not with add-ons specifically but in general. Fundamental mechanisms that the engine uses to manage its internals should not be up for disabling IMO.

**Comment by Daylily-Zeleen** [comment]
I think I grab the point that you focus about.
Maybe we should just disable generate ".uid" in "res://addons/" and enabled it for other directories.

**Comment by AThousandShips** [comment]
I don't think making UID files optional makes sense, it'd just be disabling an important and helpful feature and solves issues that people have been frustrated at for years, having them be optional just resets that and creates additional sources of bugs and frustration

Now there *should* be a proposal discussing how to handle UID files in addons, but regardless this PR needs a proposal, this shouldn't be discussed here as per the workflow guidelines, PRs are for discussing the *implementation*, proposals for discussing the idea

**Comment by Daylily-Zeleen** [comment]
This pr can be superseded by [#100994](https://github.com/godotengine/godot/pull/100994) now.

--------------------------------------
Here just talk about make the generate behavior become optional.

There have many complain in [#97352](https://github.com/godotengine/godot/pull/97352), we not disagree using uid to handle reference, we just hate it make the filesystem become mess and ugly.

**Comment by Daylily-Zeleen** [comment]
This pr can be superseded by [#100994](https://github.com/godotengine/godot/pull/100994) now.

--------------------------------------
Here just talk about make the generate behavior become optional.

There have many complain in [#97352](https://github.com/godotengine/godot/pull/97352), even though it is merged, because not all people have spare time to focus each proposal and pr, they just pull and compile by themselves, and discovery the problem at this time.
We not disagree using uid to handle reference, we just hate it make the filesystem become ugly.
For old users, we can avoid break reference by using editor's file system dock, and we don't want when we actually need manage files outside editor, files become mess.

I very agree [this comment](https://github.com/godotengine/godot/pull/97352#issuecomment-2562054587).
And I prefer [to intergrate all ".uid" into one file](https://github.com/godotengine/godot-proposals/issues/11223) and add it to vcs, although it is even less reliable.

Anyway, I know this pr can't solve all problems and satisfy everyone, just provide an possible option to make godot reach to a good future.

**Comment by elenakrittik** [comment]
> This pr can be superseded by [#100994](https://github.com/godotengine/godot/pull/100994) now.

I don't think that PR is a sufficient replacement for this PR. It only contains the "disable for addons" part of this PR, and leaves out the "disable for project" part, which is core to the proposal this PR aims to address.

**Comment by Daylily-Zeleen** [comment]
> > This pr can be superseded by [#100994](https://github.com/godotengine/godot/pull/100994) now.
> 
> I don't think that PR is a sufficient replacement for this PR. It only contains the "disable for addons" part of this PR, and leaves out the "disable for project" part, which is core to the proposal this PR aims to address.

Although the change of “disable for project” part seems more significant, the main purpose of this pr is to temporary avoid compatibility issue of plugins by disabling generate ".uid" files in "res://addons/".

**Comment by elenakrittik** [comment]
> Although the change of “disable for project” part seems more significant, the main purpose of this pr is to temporary avoid compatibility issue of plugins by disabling generate ".uid" files in "res://addons/".

I believe that's the purpose of the PR you nominated as a replacement, https://github.com/godotengine/godot/pull/100994. This PR is linked to https://github.com/godotengine/godot-proposals/issues/11571, which says "Add an option to disable generation of UID files" (i.e. disable it in general, not just for addons).

**Comment by pietru2004** [comment]
- Is there any chance that `editor/resource/generate_uid_files` could have popup when upgrading from lower godot version asking should it be enabled/disabled?
- Is there any chance that `editor/resource/generate_uid_files` could be switchable as an option when creating new project?
- Also is there chance this could be part of 4.4 before final release?

Asking cause I want to know what to expect for 4.4 regarding UID files. Basically will I be able to switch it off for my project.

**Comment by pietru2004** [comment]
Would somebody be able to tell how long review of this PR might take?
I mean only way I can realisticaly prevent `uid` files in my git repo is by adding them to git ignore of each of my projects.

**Comment by AThousandShips** [comment]
This is a new feature and we are in feature freeze until after 4.4 is released, so this can't be merged until then, but this idea itself is being discussed, see the comments on that for alternative solutions

**Comment by pietru2004** [comment]
Is this rl a new feature? To be able to disable another feature? 
Can't a setting be added during feature freeze so specified behaviour can be disabled?

**Comment by AThousandShips** [comment]
One could argue it would be part of the same feature, like the [UID upgrade tool](https://github.com/godotengine/godot/pull/102071) added recently, but this is not something that's clearly decided to be wanted or clear that it is a good solution

I don't know what the production team thinks in this case, but I'd say it'd at least require a consensus before this can be considered for 4.4, and there isn't a strong consensus if you check the reactions on this PR (15 positive, and 10 negative)

But the production team can take a look and see what they think

**Comment by AThousandShips** [comment]
> Also I had experienced a situation in the past where godot would self change uids when saving open resource.

That's something to report as a bug, not something that should be solved by not using `.uid` files IMO

**Comment by pietru2004** [comment]
> > Also I had experienced a situation in the past where godot would self change uids when saving open resource.
> 
> That's something to report as a bug, not something that should be solved by not using `.uid` files IMO

I seen it already beeing reported, I added that as just a way of saying uids never been working for me, also I belive path is easier to read than uid

**Comment by pietru2004** [comment]
@Daylily-Zeleen could this mayby target 4.4.1 - 4.4 branch ? (IDK how it works with branches, like if something get's added to master is it backported to for example 4.4.1, I assume it is not...)

**Comment by pietru2004** [comment]
Also as you said in https://github.com/godotengine/godot-proposals/discussions/11574?sort=new#discussioncomment-12379328 I tried to make modification by temp fork, but build action for some reason is crashing there (no I am not here for tech support)

**Comment by AThousandShips** [comment]
The topic of cherry picking or not is more suited to after this has been approved if it is approved, we don't normally cherry pick improvements though, but this shouldn't be restarted for 4.4.1 regardless

**Comment by pietru2004** [comment]
is there mayby info how it works? and what is cherry picking ? (I assume it is like picking 1 part of some pr instead all changes, but assumption might be wrong)

**Comment by AThousandShips** [comment]
See the [documentation](https://docs.godotengine.org/en/latest/contributing/workflow/pr_review_guidelines.html) for some details, but the approval process is based on the relevant maintainers approving the feature and deciding it's a good idea

Cherry picking is copying over a commit to another branch, you can either cherry pick parts or the whole thing, but that depends on the decision of the maintainers and production team

But if you're interested in this PR being approved you can try to get more support for it, and test it to confirm it works

**Comment by Delsin-Yu** [comment]
> could this mayby target 4.4.1 - 4.4 branch ? (IDK how it works with branches, like if something get's added to master is it backported to for example 4.4.1, I assume it is not...)

You don't need this for any version before 4.4, as we only get .uid files for 4.4

**Comment by pietru2004** [comment]
> But if you're interested in this PR being approved you can try to get more support for it, and test it to confirm it works
 I am intrested in this pr/the thing it would add, only problem I see that there were few updates in funcs this pr modifies. So it might need update if I am not mistaken. And at least my tries to make it work fail me...

**Comment by Daylily-Zeleen** [comment]
@pietru2004 just like I said last night(in my time zone), just a little modify can block generate ".uid" files, we don't need to wait next official release.

I have to clarify that just disable generate ".uid" files is not a good solution (except "res://addons/", it break something I have mentioned), UID is a useful thing, but the ".uid" files is very terrible, we need a better solution to handle UID for scripts.

**Comment by pietru2004** [comment]
I see, also I had first buuld that I think actualy compiled the editor. Gona test it when I get moment.

I see why uids can be good thing, but I don't belive it is good for them to be as single file that just stores uid, feels like import file stores more stuff...

**Comment by pietru2004** [comment]
By single file I mean single a new one for each thing meant to have UID

**Comment by pietru2004** [comment]
ok tested it and it seams to work, but needs editor reload to generate ids if variable was false previously,

Also I think logic for checking is setting anabled can be moved to `core/io/resource_loader.cpp`, into the `ResourceLoader::should_create_uid_file`

**Comment by Daylily-Zeleen** [comment]
The `ResourceLoader::should_create_uid_file()` is added later than this pr 🤣.

**Comment by pietru2004** [comment]
hm... what is difference between p_path and local_path ?

**Comment by pietru2004** [comment]
I tested artifact from checks and if file has no id (tested on gdscript) it doesn't produce errors

**Comment by 949886** [comment]
Can't wait for this feature. Need it so so so much since the dedicated files generated by the new UID system confused me a lot.

**Comment by pietru2004** [comment]
I hope this can also get into 4.4.1 and be available soon

**Comment by pietru2004** [comment]
Default value of false sounds better than needing to hand edit godot project file or reverse uid creation via git, or to get popup 'Hey do you want to use this?' or 'Hey do you want to disable this?'

**Comment by arkology** [comment]
I personally don't care about default values. Just let me to disable uid files 😭

File: core/io/resource_loader.cpp [hunk 1]
```diff
@@ -1198,6 +1198,13 @@ bool ResourceLoader::has_custom_uid_support(const String &p_path) {
 
 bool ResourceLoader::should_create_uid_file(const String &p_path) {
 	const String local_path = _validate_local_path(p_path);
+
+	bool generate_uid_files = GLOBAL_GET("editor/resource/generate_uid_files");
+	bool generate_uid_files_for_addons = GLOBAL_GET("editor/resource/generate_uid_files_for_addons");
```
Comment (MarianoGnu) [inline core/io/resource_loader.cpp:1203]: these should be per project, not per computer/editor settings
Everyone in the team should share the same setup

**Comment by pietru2004** [comment]
> I personally don't care about default values. Just let me to disable uid files 😭

+1, I hope this could be merged and it could be decided further down the line should there be popup for example and or should default be changed, I mean set default how you think is best.

Stuff like git exists and you can use git to remove .UID files, other than git there is cmd and file browser search functions, on linux I think you can use `find` command

**Comment by pietru2004** [comment]
Is there way to get somebody with write access to review this ? 
Like could somebody be pinged mayby ?

I see this
obraz
but it doesn't say liek is this in queue or something like that.


I can use modified version, but would love to have this in official release...

**Comment by pietru2004** [comment]
Also in files changed I see probably anybody can make review...
I wonder does it change anything if users supporting a change do that? Or should users without write access do it...
obraz

**Comment by elenakrittik** [comment]
> Also in files changed I see probably anybody can make review... I wonder does it change anything if users supporting a change do that? Or should users without write access do it... obraz

Everyone is free to review any PR, it's just that without special permissions approvals and/or requests for changes will not have any effect (or, in other words, will be exactly the same as choosing tjhe "Comment" option directly)

**Comment by elenakrittik** [comment]
> @Calinou think you can suggest a good reviewer for this PR? who was in charge of uid files initial implementation?

Juan

**Review by pietru2004 — APPROVED** [review]
It works... Checked it... (this be as review mode try)

**Comment by akien-mga** [comment]
I see there's a lot of activity and expectations around this PR, or in general having the possibility to disable `.uid` files in Godot 4.4, so let me clarify the project's current position.

UID support was made for a reason ([see this blog post](https://godotengine.org/article/uid-changes-coming-to-godot-4-4/)), and we didn't decide to add a new type of file on a whim. We considered and debated various options a lot and came to the conclusion that this is the lesser evil. So while we understand that some users are bothered by the new files added to the filesystem, we are not keen on adding fragmentation in the ecosystem by making it opt-out at this stage. Making UIDs optional means that everything that UIDs are meant to guarantee can no longer be achieved, and this would seriously limit what we can do in editor tooling thanks to UIDs.

We just released Godot 4.4, and we want to give it time to get proper feedback from the community, see what usability issues we can improve further with the UID support, and evaluate after some time if we need to change course, or add more flexibility.

If your main annoyance with the `.uid` files are just that there are new files that you need to check in your VCS, I would suggest that you give it a try and see if this is really that terrible in the long run. I wasn't fond myself of `.import` files when we introduced this in 3.0, and I grew used to it and have absolutely no qualms with this anymore. I expect the same will happen with `.uid` files for most users. Note that there are also more usability improvements that can be done to ease the management of `.uid` files, both in Godot and in external IDE integrations.

If you have other concerns such as actual problems caused by UID files, be sure to open proposals or bug reports about it so that it can be evaluated.

**Comment by arkology** [comment]
Thanks for the explanation! Now we finally know the official position on uid files: we are forced to have them and have to get used to it. As uid files were implemented by Juan, it is understandable and we can complain because he has vision of the project.
It makes sense to duplicate this information in the uid feedback discussion, since a significant number of comments are about disabling uid files.

**Comment by 949886** [comment]
> I see there's a lot of activity and expectations around this PR, or in general having the possibility to disable `.uid` files in Godot 4.4, so let me clarify the project's current position.
> 
> UID support was made for a reason ([see this blog post](https://godotengine.org/article/uid-changes-coming-to-godot-4-4/)), and we didn't decide to add a new type of file on a whim. We considered and debated various options a lot and came to the conclusion that this is the lesser evil. So while we understand that some users are bothered by the new files added to the filesystem, we are not keen on adding fragmentation in the ecosystem by making it opt-out at this stage. Making UIDs optional means that everything that UIDs are meant to guarantee can no longer be achieved, and this would seriously limit what we can do in editor tooling thanks to UIDs.
> 
> We just released Godot 4.4, and we want to give it time to get proper feedback from the community, see what usability issues we can improve further with the UID support, and evaluate after some time if we need to change course, or add more flexibility.
> 
> If your main annoyance with the `.uid` files are just that there are new files that you need to check in your VCS, I would suggest that you give it a try and see if this is really that terrible in the long run. I wasn't fond myself of `.import` files when we introduced this in 3.0, and I grew used to it and have absolutely no qualms with this anymore. I expect the same will happen with `.uid` files for most users. Note that there are also more usability improvements that can be done to ease the management of `.uid` files, both in Godot and in external IDE integrations.
> 
> If you have other concerns such as actual problems caused by UID files, be sure to open proposals or bug reports about it so that it can be evaluated.

I think the Godot team should hear the voice of the user instead of forcing them to accept a bunch of dedicated uid files to mess up their file system. 

There are many complaints voiced from Reddit:

https://www.reddit.com/r/godot/comments/1i1usiw/uid_changes_coming_to_godot_44/

Even some major godot youtube channel also not found of this feature:

https://youtu.be/4Q46A_UaY_4?t=788

**Comment by pietru2004** [comment]
I just gona say UIDs are the thing that always changes when you re-save resource... or at least that be a case in my projects...

**Comment by pietru2004** [comment]
I have 1 idea, we can have engine be loader and store game data in folder called data and not compile game into .pck and have engine live load assets from data, basically simple script that will load stuff from filesystem or mayby even zip

**Comment by elenakrittik** [comment]
> I see there's a lot of activity and expectations around this PR, or in general having the possibility to disable `.uid` files in Godot 4.4, so let me clarify the project's current position.
> 
> UID support was made for a reason ([see this blog post](https://godotengine.org/article/uid-changes-coming-to-godot-4-4/)), and we didn't decide to add a new type of file on a whim. We considered and debated various options a lot and came to the conclusion that this is the lesser evil. So while we understand that some users are bothered by the new files added to the filesystem, we are not keen on adding fragmentation in the ecosystem by making it opt-out at this stage. Making UIDs optional means that everything that UIDs are meant to guarantee can no longer be achieved, and this would seriously limit what we can do in editor tooling thanks to UIDs.
> 
> We just released Godot 4.4, and we want to give it time to get proper feedback from the community, see what usability issues we can improve further with the UID support, and evaluate after some time if we need to change course, or add more flexibility.
> 
> If your main annoyance with the `.uid` files are just that there are new files that you need to check in your VCS, I would suggest that you give it a try and see if this is really that terrible in the long run. I wasn't fond myself of `.import` files when we introduced this in 3.0, and I grew used to it and have absolutely no qualms with this anymore. I expect the same will happen with `.uid` files for most users. Note that there are also more usability improvements that can be done to ease the management of `.uid` files, both in Godot and in external IDE integrations.
> 
> If you have other concerns such as actual problems caused by UID files, be sure to open proposals or bug reports about it so that it can be evaluated.

To me this reads as (my words in square brackets):

---

I see there's a lot of activity and expectations around this PR, or in general having the possibility to disable `.uid` files in Godot 4.4, so let me clarify the project's current position.

Somebody else's problem is solved by this. We think that that problem is more important than the problems this solution introduced.

Besides, 4.4 just released and we're trying to hot fix as much issues that we introduced as possible without actually changing course. We first want to see if there's enough complaining to warrant rethinking our approach. [This would normally be an okay strategy. However, in this case the feedback discussion thread *alone* has TONS of feedback already given to you, most of which is not addressed by all the band-aids the project implemented to try and alleviate pain. It's already day clear what this solution fixes well and what problems it causes.]

In the meantime, just try using it. I'm sure at some point you will have been coping with the issues for so long that you will begin to accept this as a normal part of your workflow, just like with `.import` files back in the day. [The whole argument is "It felt bad, then i no longer cared enough". What else can i say.] We also have some work going on to try to fix the aforementioned issues. This work so far didn't focus on the core issue, but should help ease some of the coping. [This is essentially a re-hash of the above paragraph.]

If you have "actual" [..sorry, what???] problems with `.uid` files, feel free to file even more issues and proposals than there already are.

---

**Comment by Daylily-Zeleen** [comment]
> I just gona say UIDs are the thing that always changes when you re-save resource... or at least that be a case in my projects...

It soud like bug, you can try this in a new project and create a minimum reproduce project to submit issue.

**Comment by afreytes** [comment]
> I see there's a lot of activity and expectations around this PR, or in general having the possibility to disable `.uid` files in Godot 4.4, so let me clarify the project's current position.

This is disheartening...

**Comment by adamscott** [comment]
> I think the Godot team should hear the voice of the user instead of forcing them to accept a bunch of dedicated uid files to mess up their file system.

As @akien-mga said, we didn't just implement this system on a whim. We were in fact listening to users complaining about numerous issues linked the lack of file tracking, especially when moving files outside of Godot (which can happen a lot, even if you're cautious about it).

We recognize that it sucks, yes, but I reject the ideas we don't listen to users, which is you're suggesting.

> For example, a centralized file option for personal users, and embeded UID or dedicated UID files for the users who have collaborative needs.

A centralized file defeats the purpose of tracking files, as Godot cannot recognize that a file moved outside. Hashing the file doesn't work, because multiple files can share the same hash (when they are identical).

Embedding the UID was considered, but it's super difficult to implement on text files, as adding it can mess up the parsing of that file. It also changes the hash of the file.

**Comment by pietru2004** [comment]
> > I just gona say UIDs are the thing that always changes when you re-save resource... or at least that be a case in my projects...
> 
> It souds like bug, you can try this in a new project and create a minimum reproduce project to submit issue.

I think it was happening in 4.0 dev or beta... so some time ago, I didn't look did it happen since

**Comment by pietru2004** [comment]
Also I have spoken to few people that I don't think will look into this thread even after I gave them, link, but they also have issues with UID files and prefer to stay 4.3

**Comment by 949886** [comment]
> > I think the Godot team should hear the voice of the user instead of forcing them to accept a bunch of dedicated uid files to mess up their file system.
> 
> As @akien-mga said, we didn't just implement this system on a whim. We were in fact listening to users complaining about numerous issues linked the lack of file tracking, especially when moving files outside of Godot (which can happen a lot, even if you're cautious about it).
> 
> We recognize that it sucks, yes, but I reject the ideas we don't listen to users, which is you're suggesting.
> 
> > For example, a centralized file option for personal users, and embeded UID or dedicated UID files for the users who have collaborative needs.
> 
> A centralized file defeats the purpose of tracking files, as Godot cannot recognize that a file moved outside. Hashing the file doesn't work, because multiple files can share the same hash (when they are identical).
> 
> Embedding the UID was considered, but it's super difficult to implement on text files, as adding it can mess up the parsing of that file. It also changes the hash of the file.

It is good news to hear that the Godot team is hearing users' voice and recognized it sucks. I just hope this issue can be solved as soon as possible since it's so confusing.

---

I have another question about hash or checksum.

As long as the content of the script is the same, they will work the same way ideally. It doesn't matter to refer to which one of them, and they could be seen as the same script as well. It is ok to refer to the first one or the newest one on behalf of those duplicated scripts, isn't it? Why is the checksum still the wrong tool, as the Godot team mentioned in the blog:

> Unfortunately, checksums are the wrong tool for the job to keep track of unique file references. If you have two files with the exact same content, they would have the same checksum, but these files are still separately referenced by other scenes or resources in the project (with the expectation that these remain separate). If we relied on the file’s checksum to determine file references, then the references would break once the file contents change.

**Comment by pietru2004** [comment]
What about if there is warning near disabling UID files, like `THIS IS NOT RECOMENDED BY THE GODOT TEAM`
Can we add UID file disable switch with that and consider it as a compromise?

If not why is there need for enable alt space menu, something like auto accept quit or Main Loop type, low processor mode and many other project settings one could consider unneeded.


What I want to say by this is, some people do not want to have UID files and consider them unnecessary as it is just a file with 1 line of text and this method just disables generation of NEW UID files and the ones already generated do work. 

So please just let us select which way we want to go in our projects. Whenever we want to use UID files or whenever we want to disable their generation within our projects. Let us have this choice like you already did for many other things via [Project Settings](https://docs.godotengine.org/en/stable/classes/class_projectsettings.html).

**Comment by elenakrittik** [comment]
> What about if there is warning near disabling UID files, like `THIS IS NOT RECOMENDED BY THE GODOT TEAM` Can we add UID file disable switch with that and consider it as a compromise?

"This is not recommended but still supported" is a strange position to take, i think.

> If not why is there need for enable alt space menu, something like auto accept quit or Main Loop type, low processor mode and many other project settings one could consider unneeded.

Besides extensibility, they force the code to work within reasonable constraints by explicitly defining an API. So, it has benefits for the codebase, even if it doesn't end up being used much in the wild. I remember a good blog post about this but sadly can't find it anymore :(

**Comment by pietru2004** [comment]
@elenakrittik that all meant to just say please let us do this...

Also please remember you can disable part of godot when making stuff, I have in mind `Manage Editor Feature Profiles` panel.

In it you can remove most if not all editor features, even the stuff that could be considered core and needed for making stuff.

obraz

**Comment by MarianoGnu** [comment]
> If you have other concerns such as actual problems caused by UID files, be sure to open proposals or bug reports about it so that it can be evaluated.

@akien-mga i personally agree with you and the rest of the devs, however there's a usecase i am not sure is already considered.
When migrating code from one project to another, uids should be regenerated in the new project instead of using the old ones, because introducing foregin .uid files to an existing project makes an unlikely but reasonable chance to create duplicated uids for different files.

This proposal was trying to prevent that by disabling generation of uid files in the addons folder, because that is the main folder that is shared using the AssetLib (still not the only one)

But i believe a better solution (that i dont know if it is already implemented) is when adding files from the asset lib, to skip the uid files from the library that you are importing, and the addon maker should simply know to not rely in uids.

This can still have consequences, because tscn (and scn, tres/res) files will reference sub resources using uid, they will fix themselves on first save, but in the unlikely case that the uid already existed in the project, this file will now try to point to an unrelated file in the project.

I myself, can live with this, i want the .uid files, i think they are for the better

**Comment by pietru2004** [comment]
I wonder could disabling files UID allow for file paths like `./script.gd` in resources? 
I mean paths relative to directory an asset is in...

I find myself duplicating full directory via godot and having files from another directory be used cause UID points to it.
Like when I want to make new object in game like Machine, NPC, etc. 
Let's say I have base_machine class and I am duplicating furnace directory and want to make assembler, it end's up having furnace's script instead of script that is in it's directory. In case UID is a problem since I can have edited it  for scene to have relative and not absolute from project `res://` path.

**Comment by MarianoGnu** [comment]
> I wonder could disabling files UID allow for file paths like `./script.gd` in resources? I mean paths relative to directory an asset is in...
> 
> I find myself duplicating full directory via godot and having files from another directory be used cause UID points to it. Like when I want to make new object in game like Machine, NPC, etc. Let's say I have base_machine class and I am duplicating furnace directory and want to make assembler, it end's up having furnace's script instead of script that is in it's directory. In case UID is a problem since I can have edited it for scene to have relative and not absolute from project `res://` path.

when you duplicate a tscn file, internally subresources are mapped using first uid, then file path as fallback. the proces of duplicating he scene may not understand you also want to duplicate the subresource (because it doesn't see what other files are in the queue of files to duplicate) so it simply keeps using the old one. This is normal and was happening before, the only difference is now it happens also for scripts and not only for other resources like textures and sounds

**Comment by pietru2004** [comment]
Can we just get this feature? The option to disable UID generation?  

Preety Please.

I realize there are reasons why it was implemented, but just please let me chose do I want them or not.

**Comment by akien-mga** [comment]
@pietru2004 Please stop spamming this PR. I already explained the current status in https://github.com/godotengine/godot/pull/100973#issuecomment-2707390662.

There is no intention to merge this feature for the time being.

**Comment by pietru2004** [comment]
Ok

**Comment by akien-mga** [comment]
@MarianoGnu

> When migrating code from one project to another, uids should be regenerated in the new project instead of using the old ones, because introducing foregin .uid files to an existing project makes an unlikely but reasonable chance to create duplicated uids for different files.

I think collisions are pretty unlikely in this specific use case, but we can ensure that if such a collision happens, the editor can resolve it gracefully. Collisions can actually happen easily when duplicating files from outside Godot and that's already something we need to solve properly.

If we find that UID collisions with addons are actually happening, there are also options to make UIDs more deterministic based on the file paths, so addon UIDs wouldn't clash as they would be seeded from their `addons/<slug>/<file>` which should be unique (at least until two addons use the same slug and file names but that's a problem bigger than just UIDs).

I don't think disabling UIDs for addons specifically is a good option. Addons are one of the most relevant uses of UIDs to ensure that code can be reused independently of the filesystem structure IMO.

**Comment by MarianoGnu** [comment]
Yes, i also believe disabling uids is not the right way to go. Acceptable solutions i think of are:
- Automatically skip any .uid file when importing files from AssetLib (dont know if already done)
- Editor should print errors if detects more than 1 resource is explicitly using the same uid, this may be already happening, last time i checked the last resource was just overriding the first ome

**Comment by Daylily-Zeleen** [comment]
> I don't think disabling UIDs for addons specifically is a good option. Addons are one of the most relevant uses of UIDs to ensure that code can be reused independently of the filesystem structure IMO.

The purpose of disabling ".uid" in "res://addons/" is to **temporary** avoid UID conflict and [generate different UID in different devices if reference plugins without details](https://github.com/godotengine/godot-proposals/issues/11464#issue-2764907485).

Like I said in [#11464](https://github.com/godotengine/godot-proposals/issues/11464#issuecomment-2573235713), we can remove this limitation if we can solve these issues in future.

But it's already too late, 4.4 already release, even if we solve these problems in the future, at least plugins created by 4.4 stable are have potential UID conflict risks.


I don't care anymore, I am powerless to change anything about it.

**Comment by JoNax97** [comment]
> Yeah, it was a good run. GG Godot

Don't you think you're being a little over dramatic?

**Comment by 949886** [comment]
So many days later, maybe I should take back my words that the Godot team is hearing the voice from users.
There is no good news, not even a reply to my question from the Godot team.

These days, I always get such errors after I upgraded my project to Godot 4.4.

屏幕截图 2025-03-13 095651
屏幕截图 2025-03-13 100828

Godot 4.4 says my resource is not found, but **it actually dose exist** and it's referred by the scene.

And when I click fix Godot 4.4 says there is nothing should be fix

image

If this pr won't be approved, can the Godot team just fix these errors?

**Comment by 949886** [comment]
So many days later, maybe I should take back my words that the Godot team is hearing the voice from users.
There is no good news, not even a reply to my question from the Godot team.

These days, I always get such errors after I upgraded my project to Godot 4.4.

屏幕截图 2025-03-13 101615
屏幕截图 2025-03-13 100828

Godot 4.4 says my resource is not found, but **it actually dose exist** and it's referred by the scene.

And when I click fix Godot 4.4 says there is nothing that should be fixed

image

If this pr won't be approved, can the Godot team just fix these errors?

**Comment by akien-mga** [comment]
@949886 Please file a bug report with steps to reproduce.

**Comment by 949886** [comment]
Too tired and disappointed to rewrite an issue and care about this.
Dealing with this makes me exhausted and kills my good mood for days.
Maybe I should go back to Unity and wait for some changes.

**Comment by MarianoGnu** [comment]
@949886 your problem is not new, and doesn't seem related to uids, this happens from time to time sadly.
Godot is not good showing good errors when is unable to load a resource **because a subresource is failing to load**
In this case the resource that is failing is **fireball.tres**
if oyu open fireball.tres in a text editor and see **line 22** it will show you actually what is the resource that is missing to load because seems to be missing, but the problem is not eigther firebal.tscn, nor fireball.tres. it is a dependency

**Comment by 949886** [comment]
> @949886 your problem is not new, and doesn't seem related to uids, this happens from time to time sadly. Godot is not good showing good errors when is unable to load a resource **because a subresource is failing to load** In this case the resource that is failing is **fireball.tres** if you open fireball.tres in a text editor and see **line 22** it will show you the actual resource that is missing (or maybe another chain reaction? it never happened to me to be so nested) to load because seems to be missing, but the problem is not eigther firebal.tscn, nor fireball.tres. it is a dependency

 Thank you for your reply!

I'm just a little frastuated by this problem.
And now I think I have found what the problem is. Maybe relevant to this:

[Allow circular dependencies/references for resources in export var · Issue #7363 · godotengine/godot-proposals](https://github.com/godotengine/godot-proposals/issues/7363)

In my case. The reference is not broken; every reference is fine.
Resource `fireball.tres` and scene `fireball.tscn` have circular dependency and cause these errors.

**Comment by MarianoGnu** [comment]
Well, i am glad you found a ray of light to, at least, try to workaround the issue. Indeed circular dependencies between resources are bad, one way to prevent it is using `@export_file(tscn)` and then `load()` the scene from your fireball.tres, that will prevent the ciclic dependenci because the scene will be loaded on demand instead of on instantation (unless you run the load() on the `_init()` function, if you do that it will still try to load fireball scene while fireball scene is already trying to be loaded)

**Comment by elenakrittik** [comment]
> > Yeah, it was a good run. GG Godot
> 
> Don't you think you're being a little over dramatic?

I mean, *for me* this *is* a "GG". Not because i physically can't live with `.uid` files but because the engine is going in a direction i am simply not aligned with. Of course, i'm an extreme case, and many others will learn to cope with this (just like Akien so shamelessly suggested) because the positives of the engine outweigh the negatives for them. To repeat: that's just my view

**Comment by akien-mga** [comment]
This is going in circles and not constructive, so I'm locking this discussion.

We've heard the feedback, if we decide that we want to make UIDs optional, we'll revisit this PR.

Until then it's available for people who want to make this decision for themselves, that's the power of open source.
